crates/saya/core/src/prover/mod.rs

```diff
@@ -11,8 +19,23 @@ pub enum ProverIdentifier {
     Platinum,
 }
 
+pub async fn prove(input: String, prover: ProverIdentifier) -> anyhow::Result<String> {
+    match prover {
+        ProverIdentifier::Sharp => todo!(),
+        ProverIdentifier::Stone => prove_stone(input).await,
+        ProverIdentifier::Platinum => todo!(),
+    }
+}
+
 /// The prover client. in charge of producing the proof.
 #[async_trait]
 pub trait ProverClient {
     fn identifier() -> ProverIdentifier;
+
+    /// Setups the prover, this is a one time operation.
+    async fn setup(&self, source: &str) -> anyhow::Result<()>;
+
+    /// Generates the proof from the given trace.
+    async fn prove(&self, input: String) -> anyhow::Result<String>;
```

Comment:
Also @neotheprogramist, I do think that we should re-evaluate the fact that we combine the prover and the program that generate the trace.

Because we also need the output of the program to be able to compute the fact and do optimistic registration of the fact before the proof is actually computed and verified.

Let's discuss this in during the call, and the implication this may have.

Reply:
As we've discussed, for now we keep the execution trace generation and the proof in the same container. The output (proof) contains the information anyway.